### ISSUE TITLE
(Discussion?) Method specific model requirements?

### ISSUE THREAD

**Issue by wozniakty** [tiangolo/fastapi#600]
**Description**

This is half a question on conceptual ideas, and half a feature request maybe?

I was wondering if anyone here has any thoughts on the following scenario.

Say I'm making a profile API.  When posting a new profile, you use this model

```
class Profile(BaseModel):
  first_name: str
  last_name: str
  title: str = None
  age: int = None
  tagline: str = None
```
First name and last name are understandably required fields, because we have to have _something_ for a profile.

So now I want to add a patch endpoint, where you can just edit a single field.  For that we want to enable the following request:
```
patch service.com/profile/1234
{
  "tagline": "I am so cool"
}
```

But I haven't given a first and last name, so obviously this endpoint can't use the Profile model as its body model.

What is the best approach for this?  I can think of a variety of options, but nothing really `sparks joy.`

Obviously one approach is to just make a different model for patch.  This is certainly the most straightforward, but across a large service you might end up with a _lot_ of duplicate models.  And you'll have to manually maintain similarities between the two 🤢 

Another thought I had (not sure how easy this would be with Fast built in behaviors?  I might be missing something), was to maybe bring the method `POST v PATCH` into a field on the profile model, and write a validator that is dependent on the value of that method field...but that is definitely a bit hacky 😬 

There are likely lots of other ideas, just curious what the thoughts are from other users of this API (and or the creator)

**Comment by dmontagu** [comment]
This is a good question!

My suspicion is that there isn't a solution that works in full generality with full type-safety and uses less code than just creating separate models for each usage pattern.

You *can* get some reuse via inheritance:

```python
class Profile(BaseModel):
  first_name: str
  last_name: str
  title: str = None
  age: int = None
  tagline: str = None

class ProfilePatch(Profile):
    first_name: str = None
    last_name: str = None
```

This is typically what I do to handle this "problem" in my projects.

----

I don't think it would be too hard to make a patch-model-generating function through the use of `pydantic.create_model` that takes an input model, and returns a new model that is basically the same except all fields are optional. But mypy doesn't really like dynamically generated types, so you might have some issues there if you want to be strict.

**Comment by wozniakty** [comment]
Funny enough, I had just posted that idea as well in my work slack 😅 
![image](https://user-images.githubusercontent.com/5042313/66342641-bfe05c80-e8fe-11e9-96f9-34439d4fd318.png)

It does seem to be a more reasonable approach.

**Comment by tiangolo** [comment]
Yeah, as @dmontagu says, that's the current best approach available, using inheritance.

The thing is that you need to somehow declare some specific fields as optional in some cases, and as strictly required in others. And each case doesn't depend on the model itself but on external code (your path operation code / HTTP method used).

You could do some advanced magic tricks with `create_model`, but you'll lose editor support as type checks, completion, etc.
